lib/assets/javascripts/turbograft/turbolinks.coffee

```diff
@@ -136,39 +136,48 @@ class window.Turbolinks
 
   @loadPage: (url, xhr, options = {}) ->
     triggerEvent 'page:receive'
+    response = new TurboGraft.Response(xhr)
+    options.onlyKeys ?= []
+    options.exceptKeys ?= []
     options.updatePushState ?= true
-    if upstreamDocument = new TurboGraft.Response(xhr).document()
-      if options.partialReplace
-        reflectNewUrl url if options.updatePushState
-        updateBody(upstreamDocument, xhr, options)
-      else
-        turbohead = new TurboGraft.TurboHead(activeDocument, upstreamDocument)
-        if turbohead.hasAssetConflicts()
-          return Turbolinks.fullPageNavigate(url)
-        reflectNewUrl url if options.updatePushState
-        turbohead.waitForAssets().then((result) ->
-          updateBody(upstreamDocument, xhr, options) unless result?.isCanceled
-        )
-    else
+    options.partialReplace = isPartialReplace(options, response)
+
+    unless upstreamDocument = response.document()
       triggerEvent 'page:error', xhr
       Turbolinks.fullPageNavigate(url)
+      return
+
+    if options.partialReplace
+      reflectNewUrl url if options.updatePushState
+      updateBody(upstreamDocument, response, options)
+      return
 
-  updateBody = (upstreamDocument, xhr, options) ->
+    turbohead = new TurboGraft.TurboHead(activeDocument, upstreamDocument)
+    if turbohead.hasAssetConflicts()
+      return Turbolinks.fullPageNavigate(url)
+
+    reflectNewUrl url if options.updatePushState
```

Comment:
is it cool that `reflectNewUrl` and `reflectRedirectedUrl` can both fire? this here, and then `reflectRedirectedUrl` in the `updateBody` call?